assertion/function/assertiontree/util.go

```diff
@@ -179,10 +179,29 @@ func inverseToken(t token.Token) token.Token {
 func AddNilCheck(pass *analysis.Pass, expr ast.Expr) (trueCheck, falseCheck RootFunc, isNoop bool) {
 	noop := func(_ *RootAssertionNode) {}
 
-	binExpr, ok := astutil.Unparen(expr).(*ast.BinaryExpr)
+	expr = astutil.Unparen(expr)
+	var binExpr *ast.BinaryExpr
+
+	switch e := expr.(type) {
+	case *ast.BinaryExpr:
+		binExpr = e
+	case *ast.UnaryExpr:
+		// Check if the unary expression is a negation of a binary expression.
+		// If the unary expression encloses a nil check binary expression, then the below code interchanges the true
+		// and false branches produced by the binary expression. For example, if`!(v != nil)`, then AddNilCheck on the inner
+		// expression `(v != nil)` returns trueCheck: produceNegativeNilCheck, falseCheck: noop, isNoop = false, implying a
+		// negative nil check for the true branch. But since it is preceded with a negation (!), the below code
+		// interchanges the true and false branches, and returns trueCheck: noop, falseCheck: produceNegativeNilCheck,
+		// implying a negative nil check for the false branch.
+		if e.Op == token.NOT {
+			trueNilCheck, falseNilCheck, isNoop := AddNilCheck(pass, e.X)
+			return falseNilCheck, trueNilCheck, isNoop
+		}
+	}
```

Comment:
nit: seems simpler to get rid of the switch case and just treat the `!(expr)` as a special case and early return, right?

```go
if e, ok := expr.(*ast.UnaryExpr); ok && e.Op == token.NOT {
  return ...
}

binExpr, ok := expr
if !ok {
  return ...
}
```

Reply:
Fair enough. Updated the code to replace the `switch`.